File: 13.DFS.BFS/22.DP/22.括号生成.java
/*
 * @lc app=leetcode.cn id=22 lang=java
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
    private List<String> first = new ArrayList<>();
    private List<List<String>> G = new ArrayList<>();
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        if (n <= 0) {
            return ans;
        }

        if (G.size() == 0) {
            first.add("");
            G.add(first);
        }

        while (G.size() <= n) {
            List<String> cur = new ArrayList<>();
            for (int j = 0; j < G.size(); j++) {
                List<String> L = G.get(j);
                List<String> R = G.get(G.size() - j - 1);
                for (String l: L) {
                    for (String r: R) {
                        cur.add("(" + l + ")" + r);
                    }
                }
            }
            G.add(cur);
        }

        return G.get(n);
    }
}
// @lc code=end

File: 13.DFS.BFS/22.DP/22.括号生成.py
#
# @lc app=leetcode.cn id=22 lang=python
#
# [22] 括号生成
#

# @lc code=start


class Solution(object):
    GG = [[""]]

    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """

        G = self.GG

        if n <= 0:
            return []

        while len(G) <= n:
            cur = []
            for j in range(len(G)):
                l = G[j]
                r = G[len(G)-j-1]
                for x in l:
                    for y in r:
                        cur.append('(' + x + ')' + y)
            G.append(cur)

        return G[n]

# @lc code=end
